qcodes/instrument_drivers/ZI/ZIHDAWG8.py

```diff
@@ -0,0 +1,269 @@
+
+    def upload_waveform(self, awg_number: int, waveform: list,
+                        index: int) -> None:
+        """
+        Upload a waveform to the device memory at a given index.
+
+        Note:
+            There needs to be a place holder on the device as this only replaces
+            a data in the device memory but does not allocate new memory space.
+
+        Args:
+            awg_number: The AWG where waveform should be uploaded to.
+            waveform: An array of floating point values from -1.0 to 1.0, or
+                integers in the range (-32768...+32768)
+            index: Index of the waveform that will be replaced. If there are
+                more than 1 waveforms used then the index corresponds to the
+                position of the waveform in the Waveforms sub-tab of the AWG tab
+                in the GUI.
+        """
+        self.set('awgs_{}_waveform_index'.format(awg_number), index)
+        self.daq.sync()
+        self.set('awgs_{}_waveform_data'.format(awg_number), waveform)
+
+    def set_channel_grouping(self, group: int) -> None:
+        """
+        Set the channel grouping mode of the device.
+
+        Args:
+            group: 0: groups of 2. 1: groups of 4. 2: groups of 8 i.e., one
+                sequencer program controls 8 outputs.
+        """
+        self.set('system_awg_channelgrouping', group)
+
+    def create_parameters_from_node_tree(self, parameters: dict) -> None:
+        """
+        Create QuCoDeS parameters from the device node tree.
+
+        Args:
+            parameters: A device node tree.
+        """
+        for parameter in parameters.values():
+            getter = partial(self._getter, parameter['Node'],
+                             parameter['Type']) if 'Read' in parameter[
+                'Properties'] else None
+            setter = partial(self._setter, parameter['Node'],
+                             parameter['Type']) if 'Write' in parameter[
+                'Properties'] else False
+            options = validators.Enum(
+                *[int(val) for val in parameter['Options'].keys()]) \
+                if parameter['Type'] == 'Integer (enumerated)' else None
+            parameter_name = self._generate_parameter_name(parameter['Node'])
+            self.add_parameter(name=parameter_name,
+                               set_cmd=setter,
+                               get_cmd=getter,
+                               vals=options,
+                               docstring=parameter['Description'],
+                               unit=parameter['Unit']
+                               )
+
+    @staticmethod
+    def _generate_parameter_name(node):
+        values = node.split('/')
+        return '_'.join(values[2:]).lower()
+
+    def download_device_node_tree(self, flags: int = 0) -> dict:
+        """
+        Args:
+            flags: 0x08, Returns only nodes which are marked as setting
```

Comment:
would be nice if there was a dict that maps human-readable names to these flags, and let this method accept a list of those human-readable names. this will make the experimental code much more readable.

Reply:
Jebb, I removed it from the docstring to fit it to pep8 line length, but this is how the docstring was:

```
             flags: ziPython.ziListEnum.settingsonly -> 0x08: Returns only nodes which are marked as setting
                    ziPython.ziListEnum.streamingonly -> 0x10: Returns only streaming nodes
                    ziPython.ziListEnum.subscribedonly -> 0x20: Returns only subscribed nodes
                    ziPython.ziListEnum.basechannel -> 0x40: Return only one instance of a node in case of multiple
                    channels
                    Or any combination of flags can be used.
```
But I can add it again

Reply:
Jebb, I removed it from the docstring to fit it to pep8 line length, but this is how the docstring was:

```
             flags: ziPython.ziListEnum.settingsonly -> 0x08: Returns only nodes which are marked as setting
                    ziPython.ziListEnum.streamingonly -> 0x10: Returns only streaming nodes
                    ziPython.ziListEnum.subscribedonly -> 0x20: Returns only subscribed nodes
                    ziPython.ziListEnum.basechannel -> 0x40: Return only one instance of a node in case of multiple
                    channels
                    Or any combination of flags can be used.
```
But I can add it again

Reply:
ah, are you saying that `zhinst` (or similar) has these flags defined? in case that module is necessary for working with the instrument, then it'd be really helpful to have these `ziPython.ziListEnum.*` constants mentioned in the docstring. and in order to make it pep8 compatible, just reformat it a bit so that lines are shorted than 80 characters :)

Reply:
Yes, they are predefined constants provided by ZI. I reintroduced them to the docstring.